devtools/gradle/src/main/java/io/quarkus/gradle/builder/QuarkusModelBuilder.java

```diff
@@ -0,0 +1,299 @@
+        Set<WorkspaceModule> modules = new HashSet<>();
+        if (project.getSubprojects().isEmpty()) {
+            return Collections.singleton(getWorkspaceModule(project));
+        }
+        for (Project subproject : project.getSubprojects()) {
+            final Convention convention = subproject.getConvention();
+            JavaPluginConvention javaConvention = convention.findPlugin(JavaPluginConvention.class);
+            if (javaConvention == null) {
+                continue;
+            }
+            modules.add(getWorkspaceModule(subproject));
+        }
+        return modules;
+    }
+
+    private WorkspaceModule getWorkspaceModule(Project project) {
+        ArtifactCoords appArtifactCoords = new ArtifactCoordsImpl(project.getGroup().toString(), project.getName(),
+                project.getVersion().toString());
+        final SourceSet mainSourceSet = QuarkusGradleUtils.getSourceSet(project, SourceSet.MAIN_SOURCE_SET_NAME);
+
+        return new WorkspaceModuleImpl(appArtifactCoords, project.getProjectDir().getAbsoluteFile(),
+                project.getBuildDir().getAbsoluteFile(), getSourceSourceSet(mainSourceSet), convert(mainSourceSet));
+    }
+
+    private Set<org.gradle.api.artifacts.Dependency> getEnforcedPlatform(Project project) {
+        final Set<org.gradle.api.artifacts.Dependency> directExtension = new HashSet<>();
+        // collect enforced platforms
+        final Configuration impl = project.getConfigurations()
+                .getByName(JavaPlugin.IMPLEMENTATION_CONFIGURATION_NAME);
+        for (org.gradle.api.artifacts.Dependency d : impl.getAllDependencies()) {
+            if (!(d instanceof ModuleDependency)) {
+                continue;
+            }
+            final ModuleDependency module = (ModuleDependency) d;
+            final Category category = module.getAttributes().getAttribute(Category.CATEGORY_ATTRIBUTE);
+            if (category != null && Category.ENFORCED_PLATFORM.equals(category.getName())) {
+                directExtension.add(d);
+            }
+        }
+        return directExtension;
+    }
+
+    private Set<org.gradle.api.artifacts.Dependency> getDirectExtensionDependencies(Set<ResolvedDependency> dependencies,
+            Map<ArtifactCoords, Dependency> appDependencies) {
+        Set<org.gradle.api.artifacts.Dependency> extensions = new HashSet<>();
+        Set<ArtifactCoords> visited = new HashSet<>();
+        for (ResolvedDependency d : dependencies) {
+            ArtifactCoords key = new ArtifactCoordsImpl(d.getModuleGroup(), d.getModuleName(), "");
+            if (!visited.add(key)) {
+                continue;
+            }
+
+            Dependency appDep = appDependencies.get(key);
+            if (appDep == null) {
+                continue;
+            }
+            final org.gradle.api.artifacts.Dependency deploymentArtifact = getDeploymentArtifact(appDep);
+            if (deploymentArtifact != null) {
```

Comment:
Again, in this method you are collecting all the extensions, while we should be collecting only the first met ones. Which means if you found an extension, you should not be walking though its dependencies.

Reply:
ok, I now understant the `firstLevel` boolean, I was also missing something on the `visited`. I fixed that